File: bixlight/src/main/java/com/github/shingyx/bixlight/MainActivity.java
package com.github.shingyx.bixlight;

import android.accessibilityservice.AccessibilityServiceInfo;
import android.content.Context;
import android.content.Intent;
import android.provider.Settings;
import android.support.v7.app.AppCompatActivity;
import android.os.Bundle;
import android.text.Editable;
import android.text.TextWatcher;
import android.util.Log;
import android.view.accessibility.AccessibilityManager;
import android.widget.Button;
import android.widget.EditText;
import android.widget.TextView;

import java.util.List;
import java.util.Locale;

public class MainActivity extends AppCompatActivity {

    private static final String TAG = MainActivity.class.getSimpleName();

    private String bixlightServiceId;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        Log.v(TAG, "onCreate");
        setContentView(R.layout.activity_main);

        bixlightServiceId = String.format("%s/.%s", getPackageName(), BixlightService.class.getSimpleName());

        Button openSettings = findViewById(R.id.open_settings);
        openSettings.setOnClickListener(v -> {
            Intent intent = new Intent(Settings.ACTION_ACCESSIBILITY_SETTINGS);
            intent.addFlags(Intent.FLAG_ACTIVITY_NEW_TASK);
            startActivity(intent);
        });

        EditText throttle = findViewById(R.id.max_frequency);
        long maxRunFrequencyMs = ((BixlightApplication) getApplication()).getMaxRunFrequencyMs();
        throttle.setText(String.format(Locale.getDefault(), "%d", maxRunFrequencyMs));
        throttle.addTextChangedListener(new TextWatcher() {
            @Override
            public void beforeTextChanged(CharSequence charSequence, int start, int count, int after) {
            }

            @Override
            public void onTextChanged(CharSequence charSequence, int start, int before, int count) {
            }

            @Override
            public void afterTextChanged(Editable editable) {
                String input = editable.toString();
                long value = !input.isEmpty() ? Long.parseLong(editable.toString()) : 0;
                ((BixlightApplication) getApplication()).setMaxRunFrequencyMs(value);
            }
        });
    }

    @Override
    protected void onResume() {
        super.onResume();
        Log.v(TAG, "onResume");
        int resId = isBixlightEnabled() ? R.string.service_enabled : R.string.service_disabled;
        TextView serviceState = findViewById(R.id.serviceState);
        serviceState.setText(getText(resId));
    }

    private boolean isBixlightEnabled() {
        boolean ret = false;
        AccessibilityManager accessibilityManager = (AccessibilityManager) getSystemService(Context.ACCESSIBILITY_SERVICE);
        if (accessibilityManager != null) {
            List<AccessibilityServiceInfo> runningServices = accessibilityManager.getEnabledAccessibilityServiceList(AccessibilityServiceInfo.FEEDBACK_GENERIC);
            for (AccessibilityServiceInfo service : runningServices) {
                if (service.getId().equals(bixlightServiceId)) {
                    ret = true;
                    break;
                }
            }
        }
        Log.v(TAG, "isBixlightEnabled = " + ret);
        return ret;
    }
}

File: bixlight/src/main/java/com/github/shingyx/bixlight/MainActivity.kt
package com.github.shingyx.bixlight

import android.accessibilityservice.AccessibilityServiceInfo
import android.content.Intent
import android.os.Bundle
import android.provider.Settings
import android.text.Editable
import android.text.TextWatcher
import android.util.Log
import android.view.accessibility.AccessibilityManager
import androidx.appcompat.app.AppCompatActivity
import kotlinx.android.synthetic.main.activity_main.*

private val TAG = MainActivity::class.java.simpleName

class MainActivity : AppCompatActivity() {
    private lateinit var bixlightServiceId: String
    private lateinit var bixlightPreferences: BixlightPreferences

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        Log.v(TAG, "onCreate")
        setContentView(R.layout.activity_main)

        bixlightServiceId = "$packageName/.${BixlightService::class.java.simpleName}"
        bixlightPreferences = BixlightPreferences(this)

        openSettings.setOnClickListener {
            val intent = Intent(Settings.ACTION_ACCESSIBILITY_SETTINGS)
            intent.addFlags(Intent.FLAG_ACTIVITY_NEW_TASK)
            startActivity(intent)
        }

        maxFrequency.setText(bixlightPreferences.getSavedMaxRunFrequencyMs().toString())
        maxFrequency.addTextChangedListener(object : TextWatcher {
            override fun beforeTextChanged(charSequence: CharSequence, start: Int, count: Int, after: Int) {
            }

            override fun onTextChanged(charSequence: CharSequence, start: Int, before: Int, count: Int) {
            }

            override fun afterTextChanged(editable: Editable) {
                val input = editable.toString()
                val value = if (input.isNotEmpty()) Integer.parseInt(input) else 0
                bixlightPreferences.saveMaxRunFrequencyMs(value)
            }
        })
    }

    override fun onResume() {
        super.onResume()
        Log.v(TAG, "onResume")
        val resId = if (isBixlightEnabled()) R.string.service_enabled else R.string.service_disabled
        serviceState.text = getText(resId)
    }

    private fun isBixlightEnabled(): Boolean {
        var ret = false
        val accessibilityManager = getSystemService(AccessibilityManager::class.java)
        if (accessibilityManager != null) {
            val runningServices = accessibilityManager.getEnabledAccessibilityServiceList(AccessibilityServiceInfo.FEEDBACK_GENERIC)
            ret = runningServices.any { it.id == bixlightServiceId }
        }
        Log.v(TAG, "isBixlightEnabled = $ret")
        return ret
    }
}
